### ISSUE TITLE
Augmentations in training loop?

### ISSUE THREAD

**Issue by choosehappy** [choosehappy/PatchSorter#34]
@jacksonjacobs1 , in looking here:

https://github.com/choosehappy/PatchSorter/blob/9771eb52c78eda735ec6b01ef8dc6e3b1d89d2fd/patchsorter/dl/training.py#L361

it seems like the main training loop is doing the augmentations instead of the dataloader?

**Comment by jacksonjacobs1** [comment]
Thanks for pointing this out. I'll open a PR shortly in which
1. ShardDataset implements IterableDataset
2. A dataloader is added
3. `worker_init_fn` passed to dataloader, as is done in QuickAnnotator to ensure safe database connection pooling across processes
https://github.com/choosehappy/QuickAnnotator/blob/73190e35431201409b98276d8619eb70b7595c76/quickannotator/dl/training.py#L120


In the current implementation, each ray worker "owns" the subset of locally available shard ids, mod'ed by the ray worker's local rank:
https://github.com/choosehappy/PatchSorter/blob/9771eb52c78eda735ec6b01ef8dc6e3b1d89d2fd/patchsorter/dl/training.py#L647

BTW I just noticed a bug in this function - we should modulo the index of the local shard list rather than the shard ids themselves, since shard_id are not sequential integers.
```python
return [s for i, s in enumerate(shard_ids) if i % num_local_workers == rank]
```
Rather than
```python
return [s for s in shard_ids if s % num_local_workers == rank]
```


## Implementation of dataloader shard splitting
I'm currently thinking that dataloader workers should be assigned mutually exclusive shard subsets, rather than assigning each worker a mutually exclusive subsets of batches within each shard. Do you agree? 

If so, the implementation would look like the following:

After passing shard assignments to each ray train worker,  perform a second shard assignment at the dataloader level, i.e., compute the following in the `worker_init_fn`:
```python
compute_shard_assignments(ray_worker_shard_ids, num_dataloader_workers, dataloader_worker_rank)
```

The efficiency gain hinges on intelligently choosing `num_ray_workers`, `num_dataloader_workers`, `global_num_shards`

Say we have the following setup:
- 4 total num_ray_workers
- 8 dataloader workers per ray worker
- 32 global citus shards

Then the global citus shards get whittled down very quickly:
32 * (1/4) * (1/8) = 1

The key constraint in this setup is the number of dataloader workers must be <= the number of shards assigned to the ray worker, otherwise some workers will be assigned 0 shards.

**Comment by choosehappy** [comment]
> I'm currently thinking that dataloader workers should be assigned mutually exclusive shard subsets, rather than assigning each worker a mutually exclusive subsets of batches within each shard. Do you agree?

yes, i think that is the case - it should make things a bit cleaner and stable. it should also result in better gradient "mixing", since each shard is more likely to have more diverse data, so we'd want more shards contributing to each back prop


> The key constraint in this setup is the number of dataloader workers must be <= the number of shards assigned to the ray worker, otherwise some workers will be assigned 0 shards.

hmmm this makes things a bit more messy - how does this scale across different configurations? i.e., in small/medium/lerge cohort sizes, what is the likeihood we end up with workers doing with 0 shard?

in a more sophitocated version, we would dedicate a function to intelligently figuring this out on a per-use case basis - but this is out of scope for this version. i think we instead need to pick what will be the most robust out of the box for the all uses, and then see if there is relaly value in optimizing later?

**Comment by jacksonjacobs1** [comment]
> hmmm this makes things a bit more messy - how does this scale across different configurations? i.e., in small/medium/lerge cohort sizes, what is the likeihood we end up with workers doing with 0 shard? in a more sophitocated version, we would dedicate a function to intelligently figuring this out on a per-use case basis - but this is out of scope for this version. i think we instead need to pick what will be the most robust out of the box for the all uses, and then see if there is relaly value in optimizing later?

Looking at the smallest recommended use case of PatchSorter, say 10,000 objects, it would be silly to use distributed training, so I'll set 
- 1 training worker
- 8 dataloading workers
- 32 shards
32 * (1/8) = 4 shards per dataloader worker. 

However, I would not recommend 32 shards for this use case as each shard would only have 312 objects, much smaller than a batch. Shards which do not evenly divide into batches result in "wasted" IO round trips. If we had 8 shards, we would still have 1 shard per dataloader worker, and each shard would have ~1250 objects. If the dataset's db query paginates by 1024 objects, one of every two db queries would produce an undersized page of objects.

Wasted IO trips could be prevented by having each db query UNION across shard i and shard i + 1, but this adds complexity (need to keep track of 2 cursors rather than 1)

I also note that in the final implementation the batch size will be set in the dataloader, and will thus be completely abstracted away from the database page size. So the model will not receive undersized batches that could interfere with losses.

**Comment by choosehappy** [comment]
currently as well we're doing drop_last_batch if not full - which is the larger looming concern- since if we end up with a shard size that is < batch_size, i am actually unsure if any learning takes place at all, which seems like a very bad edge case to be in

we also essentially don't know the # of objects from the onset (which could change later on if the user uploads more), _as well_ they could spin up more compute if they have a large job and while the system is set up (e.g., citus) to redistirubte the shards (is everything else?) - the DL config has to be robust to that dynamic scaling

so in the end, what is the most robust option that we can set or how do we manage this more...fluid..situation?

**Comment by jacksonjacobs1** [comment]
As of now, we don't have drop last batch behavior implemented anywhere

Redistributing shards is relatively straightforward with citus - it won't cause any problems with aggregation tables or training state (e.g., last_cursor_location), since this distributed state is tightly coupled at the shard level via co-location.

changing the number of shards is not as straightforward because it would likely require aggregations and training state to be thrown away. This is the function that citus provides for changing shard count:
https://docs.citusdata.com/en/stable/develop/api_udf.html#alter-distributed-table

**Comment by choosehappy** [comment]
>As of now, we don't have drop last batch behavior implemented anywhere

this would come in when moving this to a dataloader (e.g., is it in my research code)

otherwise, we'll need to add a lot more debugging code and still potentially skip

e.g., many of the losses are based on the idea that there are multiple points in order to "pull" and "push" apart batches (e.g., how simclr works) - in a worse case, if there is only 1 patch in the final batch - i don't know what the consequence would be, but i suspect a number of the losses would fail entirely and throw errors due to mishapen data


i definitely think we want to avoid changing the shard count - if there is a large amount of data (E.g., TB or PB), i don't know what the cost would be, but i imagine it would immense in terms of both time and storage (i suspect its an atomic copy->delete operation as opposed to a row by row move)


where does this leave us?

**Comment by jacksonjacobs1** [comment]
Our first priority is that every patch receives a prediction, so the dataloader must not drop the last batch for that reason. Unless you see some other more elegant way, I think we should just add protection against loss computation failures - the most naive solution would just be to avoid backprop if the batch is undersized.

I agree that the shard count should be fixed once set. I also think that it should be set in constants.py (devops-facing rather than user-facing). Do you agree?

**Comment by choosehappy** [comment]
> Our first priority is that every patch receives a prediction, so the dataloader must not drop the last batch for that reason

ah, thats a very fair point - please make sure thats noted down somewhere : ) 

> the most naive solution would just be to avoid backprop if the batch is undersized.

this is an interesting idea - maybe it can be further extended to add robustness. e.g., don't do backprop if any of the associated losses (or total loss) is NaN?  i think we had something similar in place already, setting individual losses to 0 if they're NaN - or maybe that is already sufficient in this use case?  ultimately if we +/- a batch here or there it won't change anything - the more important thing is to make sure that there is no backprop that puts the system in an unsane situation

> I also think that it should be set in constants.py

yes agree
